### PR TITLE
Fix to support no-batch-dim inputs in ConvTransposeNd._output_padding

### LINKED ISSUE #75889: forward() method of ConvTranspose2d does not respect output_size parameter
### 🐛 Describe the bug

The deconvolution module torch.nn.ConvTranspose2d accepts in its forward() method an optional parameter output_size for specifying the shape of the output. With this, an output padding is computed such that the given shape is produced by the deconvolution. If this is not possible, a respective ValueError is raised.
However, in the following code example an output with a different shape (1, 232, 200) than requested (1, 240, 200) is produced.
(The check fails for i > 0.) Curiously, it works with 4-dimensional input tensors, e.g. with shape (3, 1, 15, 13).

```python
import torch
import torch.nn as nn

i = 1
in_shapes = [(1, 30, 25), (1, 15, 13), (1, 8, 7), (1, 4, 4), (1, 2, 2)]
in_shape = in_shapes[i]
out_shape = (1, 240, 200)
tensor = torch.zeros(in_shape)
up = 8 * 2 ** i
pad = up // 2 - 1
deconv = nn.ConvTranspose2d(1, 1, kernel_size=up, stride=up, padding=pad, bias=False)
assert deconv(tensor, output_size=out_shape).shape == out_shape
```

### Versions

```
PyTorch version: 1.11.0
Is debug build: False
CUDA used to build PyTorch: Could not collect
ROCM used to build PyTorch: N/A

OS: Microsoft Windows 10 Education
GCC version: Could not collect
Clang version: Could not collect
CMake version: Could not collect
Libc version: N/A

Python version: 3.8.12 (default, Oct 12 2021, 03:01:40) [MSC v.1916 64 bit (AMD64)] (64-bit runtime)
Python platform: Windows-10-10.0.19043-SP0
Is CUDA available: False
CUDA runtime version: 11.3.109
GPU models and configuration: GPU 0: NVIDIA GeForce RTX 2080 Ti
Nvidia driver version: 497.09
cuDNN version: C:\Program Files\NVIDIA GPU Computing Toolkit\CUDA\v11.3\bin\cudnn_ops_train64_8.dll
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

Versions of relevant libraries:
[pip3] numpy==1.21.5
[pip3] torch==1.11.0
[pip3] torchaudio==0.11.0
[pip3] torchvision==0.12.0
[conda] blas                      1.0                         mkl
[conda] cudatoolkit               11.0.221             h74a9793_0
[conda] mkl                       2021.4.0           haa95532_640
[conda] mkl-service               2.4.0            py38h2bbff1b_0
[conda] mkl_fft                   1.3.1            py38h277e83a_0
[conda] mkl_random                1.2.2            py38hf11a4ad_0
[conda] numpy                     1.21.5           py38ha4e8547_0
[conda] numpy-base                1.21.5           py38hc2deb75_0
[conda] pytorch                   1.11.0              py3.8_cpu_0    pytorch
[conda] pytorch-mutex             1.0                         cpu    pytorch
[conda] torchaudio                0.11.0                 py38_cpu    pytorch
[conda] torchvision               0.12.0                 py38_cpu    pytorch
```

cc @ezyang @gchanan @zou3519 @albanD @mruberry @jbschlosser @walterddr @kshitij12345

### PR DESCRIPTION
**Pull request by jbschlosser**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #76151

Fixes #75889

### CODE REVIEW

File: torch/nn/modules/conv.py [hunk 1]
```diff
@@ -610,22 +610,24 @@ def __init__(self, in_channels, out_channels, kernel_size, stride,
...
+                        num_spatial_dims: int, dilation: Optional[List[int]] = None) -> List[int]:
         if output_size is None:
             ret = _single(self.output_padding)  # converting to list if was not already
         else:
-            k = input.dim() - 2
-            if len(output_size) == k + 2:
-                output_size = output_size[2:]
+            has_batch_dim = input.dim() == num_spatial_dims + 2
+            num_non_spatial_dims = 2 if has_batch_dim else 1
+            k = num_spatial_dims
```
Comment (albanD) [inline torch/nn/modules/conv.py:619]: Why use `k` here? Can't we just keep using `num_spatial_dims` ?

**Review by albanD — APPROVED** [review]
Minor naming thing, LGTM otherwise

File: torch/nn/modules/conv.py (hunk 1 above)
Comment (jbschlosser) [reply torch/nn/modules/conv.py:619]: You're right :) I was being lazy to minimize changes. Will fix

**Review by albanD — APPROVED** [review]
A lot more readable! Thanks!
